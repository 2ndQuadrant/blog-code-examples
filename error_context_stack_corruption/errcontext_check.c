#include "postgres.h"
#include "errcontext_check.h"

/*
 * Example code for blog article https://www.2ndquadrant.com/wp-admin/post.php?post=58103
 */

/*
 * Check that error context stack entries don't leak if we escape scope.
 *
 * Implementation of pgl_errcontext_check()
 */
void
check_errcontext_stack_on_return(const ErrorContextCallback * const cb)
{
    ErrorContextCallback *stack;

    /*
     * No reference to the stack-allocated ErrorContextCallback that's leaving
     * scope may remain on the global error_context_stack.
     *
     * There doesn't seem to be a good lightweight way to acquire the caller
     * function name; we can't use clang's __builtin_FUNCTION() etc because we
     * can't supply default-parameters.
     */
    for (stack = error_context_stack; stack != NULL; stack = stack->previous)
    {
        if (stack == cb)
        {
            /*
             * Fix the error context stack and ERROR the backend. If the postgres
             * version supports it, emit a backtrace for the error to the log
             * as well.
             *
             * A more useful implementation for real world use would PANIC here
             * or would use pgl_errcontext_check_cassert() and would then
             * Assert(stack != cb). But in this case we want to keep the backend
             * alive, so we don't have to restart it.
             */
            error_context_stack = stack->previous;
            ereport(ERROR,
                    (errcode(ERRCODE_INTERNAL_ERROR),
                     errmsg_internal("leaked an errcontext callback pointer"),
                     errbacktrace()));
        }
    }
}
