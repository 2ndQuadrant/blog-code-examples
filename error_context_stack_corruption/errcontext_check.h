#ifndef ERRCONTEXT_CHECK_H
#define ERRCONTEXT_CHECK_H

/*
 * Example code for blog article https://www.2ndquadrant.com/wp-admin/post.php?post=58103
 */

/*
 * gcc/clang/icc feature detection support
 *
 * https://clang.llvm.org/docs/LanguageExtensions.html
 * https://gcc.gnu.org/onlinedocs/cpp/_005f_005fhas_005fattribute.html
 */
#ifndef __has_attribute
  #define __has_attribute(x) 0
#endif

/*
 * pg_attribute_cleanup(cbfunc) can be used to annotate a stack (auto) variable
 * with a cleanup call that's automatically executed by the compiler on any
 * path where it leaves scope.
 *
 * This is useful for asserting that something got cleaned up.
 *
 * To allow for checks to run only on cassert, a variant
 * pgl_attribute_cleanup_cassert(cb) is provided that's only defined when using
 * --enable-cassert builds.
 *
 * Ignored when unsupported, so you can only use this to make sure that some
 * expected cleanup task happened before the variable exited the current scope,
 * you can't do anything destructor-like with it.
 *
 * This has been in gcc since forever, so we don't need any meaningful version
 * guard. clang supports it too.
 *
 * See https://gcc.gnu.org/onlinedocs/gcc/Common-Variable-Attributes.html#index-cleanup-variable-attribute
 */
#if __has_attribute(cleanup)
#define pgl_attribute_cleanup(cb) __attribute__((cleanup(cb)))
#ifdef USE_ASSERT_CHECKING
#define pgl_attribute_cleanup_cassert(cb) __attribute__((cleanup(cb)))
#else
#define pgl_attribute_cleanup_cassert(cb)
#endif
#else
#define pgl_attribute_cleanup(cb)
#define pgl_attribute_cleanup_cassert(cb)
#endif

extern void check_errcontext_stack_on_return(const ErrorContextCallback * const cb);

/*
 * Use as
 *
 *     ErrorContextCallback ctx pgl_errcontext_check();
 *
 * to auto-check the entry is no longer referenced on the stack at end-of-scope.
 *
 * Doesn't check longjmp() so won't treat elog(ERROR) as a problem, but we
 * unwind automatically in that case so it's fine.
 *
 * Enabled here for both cassert and non-cassert builds because most people
 * trying out this extension won't have access to a cassert build.
 */
#define pgl_errcontext_check() pgl_attribute_cleanup(check_errcontext_stack_on_return)

/*
 * Backwards compatibility to allow errbacktrace() to be used on older postgres
 */
#if PG_VERSION_NUM < 130000 && !defined(PG_LOG_BACKTRACE)
#define errbacktrace() 0
#endif

#endif
