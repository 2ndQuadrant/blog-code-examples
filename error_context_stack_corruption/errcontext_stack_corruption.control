comment = 'Error context stack corruption examples'
default_version = '1.0'
schema = 'errcontext_stack_corruption'
module_pathname = '$libdir/errcontext_stack_corruption'
relocatable = false
