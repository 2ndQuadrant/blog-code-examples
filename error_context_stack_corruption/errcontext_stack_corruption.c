/*
 * Example code for blog article
 *
 * Craig Ringer <craig.ringer@2ndquadrant.com>
 * https://www.2ndquadrant.com/en/blog/dev-corner-error-context-stack-corruption/
 */

#include "postgres.h"

#include "funcapi.h"
#include "utils/builtins.h"

#include "errcontext_check.h"

PG_MODULE_MAGIC;

/*
 * Dummy code
 */
static void
do_the_thing(void)
{
    ereport(INFO,
            (errmsg("did the thing")));
};

/*
 * Example 1
 */
static void
my_func(bool do_it)
{
    if (!do_it)
    {
        elog(WARNING, "not doing it!");
        return;
    }

    do_the_thing();
}

/*
 * Example 2
 */
struct my_func_ctx_arg
{
    bool do_it;
};

static void
my_func_ctx_callback(void *arg)
{
    struct my_func_ctx_arg *ctx_arg = arg;
    errcontext("during my_func(do_it=%d)", ctx_arg->do_it);
}

static ErrorContextCallback *leaked_errcontext_ptr;

/*
 * WARNING WARNING WARNING
 *
 * This code is deliberately BUGGY. DO NOT COPY IT.
 */
static void
my_func_with_errcontext_BUGGY(bool do_it)
{
    ErrorContextCallback myerrcontext;
    struct my_func_ctx_arg ctxinfo;

    ctxinfo.do_it = do_it;
    myerrcontext.callback = my_func_ctx_callback;
    myerrcontext.arg = &ctxinfo;
    myerrcontext.previous = error_context_stack;
    error_context_stack = &myerrcontext;

    if (!do_it)
    {
        /*
         * -----------------------
         * WARNING WARNING WARNING
         * -----------------------
         *
         * This code is deliberately BUGGY. DO NOT COPY IT.
         */
        elog(LOG, "leaking error context stack pointer %p", error_context_stack);
        leaked_errcontext_ptr = error_context_stack;
        elog(WARNING, "not doing it and leaking the error context callback pointer!");
        return;
    }

    do_the_thing();

    /*
     * Note that this isn't necessarily strictly correct for complex functions
     * that may involve transaction management, etc:
     */
    Assert(error_context_stack == &myerrcontext);
    error_context_stack = myerrcontext.previous;
}

static void
my_func_with_errcontext_fixed(bool do_it)
{
    ErrorContextCallback myerrcontext;
    struct my_func_ctx_arg ctxinfo;

    ctxinfo.do_it = do_it;
    myerrcontext.callback = my_func_ctx_callback;
    myerrcontext.arg = &ctxinfo;
    myerrcontext.previous = error_context_stack;
    error_context_stack = &myerrcontext;

    if (!do_it)
    {
        elog(INFO, "not doing it (correctly)");
        /* don't shortcut 'return', fall through */
    }
    else
    {
        do_the_thing();
    }

    /* Safer way to unwind errcontext stack */
    if (error_context_stack == &myerrcontext)
        error_context_stack = myerrcontext.previous;
}

/*
 * WARNING WARNING WARNING
 *
 * This code is deliberately BUGGY. DO NOT COPY IT.
 *
 * This modifies my_func_with_errcontext_BUGGY very minimally to
 * add the pgl_errcontext_check() attribute on the declaration
 * of the ErrorContextCallback variable.
 *
 * It will fix the problem and ERROR when called, preventing a crash, but only
 * if the compiler supports __attribute__((cleanup)).
 */
static void
my_func_with_errcontext_ERRDETECT(bool do_it)
{
    ErrorContextCallback myerrcontext pgl_errcontext_check();
    struct my_func_ctx_arg ctxinfo;

    ctxinfo.do_it = do_it;
    myerrcontext.callback = my_func_ctx_callback;
    myerrcontext.arg = &ctxinfo;
    myerrcontext.previous = error_context_stack;
    error_context_stack = &myerrcontext;

    if (!do_it)
    {
        /*
         * -----------------------
         * WARNING WARNING WARNING
         * -----------------------
         *
         * This code is deliberately BUGGY. DO NOT COPY IT.
         */
        elog(LOG, "leaking error context stack pointer %p", error_context_stack);
        leaked_errcontext_ptr = error_context_stack;
        elog(WARNING, "not doing it and leaking the error context callback pointer!");
        return;
    }

    do_the_thing();

    /*
     * Note that this isn't necessarily strictly correct for complex functions
     * that may involve transaction management, etc:
     */
    if (error_context_stack == &myerrcontext)
        error_context_stack = myerrcontext.previous;
}

/*
 * Wrappers to make the examples SQL-callable and more reliably trigger the
 * crash.
 *
 * They rely on attributes to control compilation and prevent inlining, but
 * you can probably just use -O0 instead.
 */

#if __has_attribute(noinline)
#define ext_noinline() __attribute__((noinline))
#else
#define ext_noinline()
#endif
#if __has_attribute(optimize)
#define ext_noopt() __attribute__((optimize(0)))
#else
#define ext_noopt()
#endif

ext_noinline() ext_noopt() static void
clobber_the_stack(void)
{
    /*
     * Write a chunk of nonsense into the stack so that any space for the
     * error_context_stack gets overwritten by nonsense.
     */
#define N_DUMMY_BYTES 2000
    char dummy_bytes[N_DUMMY_BYTES];
    memset(&dummy_bytes[0], '\x7f', N_DUMMY_BYTES);
#undef N_DUMMY_BYTES
}

ext_noinline() ext_noopt() static void
call_with_padded_stack(bool do_it, const char *variant)
{
    /*
     * We pad the stack with some nonsense, but not as much as will
     * get written by the clobber_the_stack() call. This ensures that when
     * we call ereport(), the stack for errstart/errfinish don't overwrite
     * the stack space our eventual pointer into error_context_stack gets
     * put in.
     */
#define N_DUMMY_BYTES 1000
    char dummy_bytes[N_DUMMY_BYTES];
    memset(&dummy_bytes[0], '\x6f', N_DUMMY_BYTES);
#undef N_DUMMY_BYTES

    if (strcmp(variant, "simple") == 0)
        my_func(do_it);
    else if (strcmp(variant, "errcontext_buggy") == 0)
        my_func_with_errcontext_BUGGY(do_it);
    else if (strcmp(variant, "errcontext_fixed") == 0)
        my_func_with_errcontext_fixed(do_it);
    else if (strcmp(variant, "errcontext_detect") == 0)
        my_func_with_errcontext_ERRDETECT(do_it);
    else
        ereport(ERROR,
                (errmsg("unrecognised \"variant\" argument: %s", variant)));
}

PG_FUNCTION_INFO_V1(my_func_sql);

Datum
my_func_sql(PG_FUNCTION_ARGS)
{
    bool do_it = PG_GETARG_BOOL(0);
    const char * variant = text_to_cstring(PG_GETARG_TEXT_P(1));

    leaked_errcontext_ptr = NULL;

    call_with_padded_stack(do_it, variant);
    clobber_the_stack();

    ereport(INFO,
            (errmsg("after return from requested call")));

    leaked_errcontext_ptr = NULL;

    PG_RETURN_VOID();
}
