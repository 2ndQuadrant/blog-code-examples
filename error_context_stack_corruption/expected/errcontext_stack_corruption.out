CREATE EXTENSION errcontext_stack_corruption;
ERROR:  parameter "schema" cannot be specified when "relocatable" is true
SET client_min_messages = notice;
SELECT my_func(1, 'simple');
ERROR:  function my_func(integer, unknown) does not exist
LINE 1: SELECT my_func(1, 'simple');
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT my_func(0, 'simple');
ERROR:  function my_func(integer, unknown) does not exist
LINE 1: SELECT my_func(0, 'simple');
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT my_func(1, 'errcontext_fixed');
ERROR:  function my_func(integer, unknown) does not exist
LINE 1: SELECT my_func(1, 'errcontext_fixed');
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT my_func(0, 'errcontext_fixed');
ERROR:  function my_func(integer, unknown) does not exist
LINE 1: SELECT my_func(0, 'errcontext_fixed');
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
